sound/soc/sof/compress.c

```diff
@@ -296,30 +296,40 @@ static int sof_compr_trigger(struct snd_soc_component *component,
 				  &reply, sizeof(reply));
 }
 
-static int sof_compr_copy(struct snd_soc_component *component,
-			  struct snd_compr_stream *cstream,
-			  char __user *buf, size_t count)
+static int sof_compr_copy_playback(struct snd_compr_runtime *rtd,
+				   char __user *buf, size_t count)
 {
-	struct snd_compr_runtime *rtd = cstream->runtime;
-	unsigned int offset, n;
 	void *ptr;
-	int ret;
-
-	if (count > rtd->buffer_size)
-		count = rtd->buffer_size;
+	unsigned int offset, n;
 
 	div_u64_rem(rtd->total_bytes_available, rtd->buffer_size, &offset);
 	ptr = rtd->dma_area + offset;
 	n = rtd->buffer_size - offset;
 
 	if (count < n) {
-		ret = copy_from_user(ptr, buf, count);
+		if (copy_from_user(ptr, buf, count))
+			return -EFAULT;
 	} else {
-		ret = copy_from_user(ptr, buf, n);
-		ret += copy_from_user(rtd->dma_area, buf + n, count - n);
+		if (copy_from_user(ptr, buf, count))
+			return -EFAULT;
+
+		if (copy_from_user(rtd->dma_area, buf + n, count - n))
```

Comment:
I don't get why we moved from accumulating the bytes received to testing from something different from zero. Even if this was desired, it would need to be done in a separate patch prior to the introduction of a new function

Reply:
I have to agree with Pierre here, I can also see some changes that are unsound... I need a proper explanation before ever accepting them.

Reply:
The reason I changed this is because I believe `sof_compr_copy` should return something negative in case of error so compress_offload API can know that the copy operation went wrong. From my understanding, in the previous version, `sof_compr_copy` would return something >= 0 even if the copy_{from/to}_user failed. You're right though, this needs to be added to a separate patch.

Reply:
@plbossart marking this as resolved after @LaurentiuM1234 last comment. We revert back to only move code around and add capture.